pkg/volume/volume.go

```diff
@@ -237,49 +236,83 @@ func NewParameters(params map[string]string) (Parameters, error) {
 }
 
 // Convert parameters into a modify-object that reconciles any differences between parameters and resource group
-func (params *Parameters) ToResourceGroupModify(rg *lapi.ResourceGroup) (lapi.ResourceGroupModify, error) {
+func (params *Parameters) ToResourceGroupModify(rg *lapi.ResourceGroup) (lapi.ResourceGroupModify, bool) {
+	changed := false
 	rgModify := lapi.ResourceGroupModify{
 		OverrideProps: make(map[string]string),
 	}
 
-	if rg.Props == nil {
-		rg.Props = make(map[string]string)
+	if rg.SelectFilter.PlaceCount != params.PlacementCount {
+		changed = true
+		rgModify.SelectFilter.PlaceCount = params.PlacementCount
 	}
 
-	rgModify.SelectFilter.PlaceCount = params.PlacementCount
+	// LINSTOR does not accept "" as a placeholder for an absent storage pool, so we must take care to never try to
+	// set it in such cases.
+	if params.StoragePool != "" {
+		if rg.SelectFilter.StoragePool != params.StoragePool {
+			changed = true
+			rgModify.SelectFilter.StoragePool = params.StoragePool
+		}
+
+		existingPool, ok := rg.Props[lc.KeyStorPoolName]
+		if !ok || existingPool != params.StoragePool {
+			changed = true
+			rgModify.OverrideProps[lc.KeyStorPoolName] = params.StoragePool
+		}
+	}
+
+	if !reflect.DeepEqual(rg.SelectFilter.ReplicasOnSame, params.ReplicasOnSame) {
+		changed = true
+		rgModify.SelectFilter.ReplicasOnSame = params.ReplicasOnSame
+	}
 
-	if params.StoragePool != "" { // otherwise we set the storagepool to "", which does not make LINSTOR happy
-		rgModify.SelectFilter.StoragePool = params.StoragePool
-		rgModify.OverrideProps[lc.KeyStorPoolName] = params.StoragePool
+	if !reflect.DeepEqual(rg.SelectFilter.ReplicasOnDifferent, params.ReplicasOnDifferent) {
```

Comment:
besides my some general dislike of `reflect`, is this really necessary? these are just string slices IIRC. also it could result in `updated = true` if only the order (but not the semantic) changes, right? all in all I think I would use simple helper function checking for `len()` and if every x is in y.

Reply:
:+1: added a simple helper function. While order is not important for `replicasOnSame` and `replicasOnDifferent`, it can be important for the layerstack. In any case, these values shouldn't even change, so I think its fine to assume they are always in the same order